# Lab2.1: Words, concepts, semantic relations in Wordnet-NLTK

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

In this notebook, you are going to work with wordnet databases that have been incorporated in the NLTK package.
Detailed information how to access and use wordnet can be found here: http://www.nltk.org/howto/wordnet.html

Study the documentation and make yourself familiar with the different commands. Some of documentation is also explained below.

In [1]:
from nltk.corpus import wordnet as wn

Look up a word using the "wn.synsets()" function. This will give you a list of synsets in which the lookup string is matched with a lemma (synonym).

In [6]:
all_dog_synsets = wn.synsets('dog')
print('Number of synsets with "dog" as a synonym:', len(all_dog_synsets))
print(all_dog_synsets)

Number of synsets with "dog" as a synonym: 8
[Synset('dog.n.01'), Synset('frump.n.01'), Synset('dog.n.03'), Synset('cad.n.01'), Synset('frank.n.02'), Synset('pawl.n.01'), Synset('andiron.n.01'), Synset('chase.v.01')]


Note that the synsets in this list are printed by listing the first synonym only. Also note that we got nouns and verbs. We can iterate over the list to get each synset as an 'object' and next call specific functions for each synset:

In [4]:
for synset in all_dog_synsets:
    print()
    print('The synset =', synset)
    print('Type', type(synset))
    print('The synonyms = ', synset.lemmas())
    print('The definition =', synset.definition())
    print('The full path of hypernyms =', synset.hypernym_paths())
    print('The maximum depth of its hyponymy chain is = ', synset.max_depth())



The synset = Synset('cat.n.01')
Type <class 'nltk.corpus.reader.wordnet.Synset'>
The synonyms =  [Lemma('cat.n.01.cat'), Lemma('cat.n.01.true_cat')]
The definition = feline mammal usually having thick soft fur and no ability to roar: domestic cats; wildcats
The full path of hypernyms = [[Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('chordate.n.01'), Synset('vertebrate.n.01'), Synset('mammal.n.01'), Synset('placental.n.01'), Synset('carnivore.n.01'), Synset('feline.n.01'), Synset('cat.n.01')]]
The maximum depth of its hyponymy chain is =  13

The synset = Synset('guy.n.01')
Type <class 'nltk.corpus.reader.wordnet.Synset'>
The synonyms =  [Lemma('guy.n.01.guy'), Lemma('guy.n.01.cat'), Lemma('guy.n.01.hombre'), Lemma('guy.n.01.bozo')]
The definition = an informal term for a youth or man
The full path of hypernyms = [[Synset('entity.n.01'), Synset('phy

You can also obtain synsets with a certain part-of-speech only by passing a part-of-speech value as a parameter.

In [9]:
all_dog_verb_synsets = wn.synsets('dog', 'v')
print(all_dog_verb_synsets)

[Synset('chase.v.01')]


Various functions can be called on the synset object yielding different data structures. Try some to get a feeling for it. Below we show the specific relations that synsets can have besides hyponymy:

In [10]:
doggy_synset = all_dog_synsets[0]
#### Part - to -  whole relations:
print('Part holonyms:',doggy_synset.part_holonyms())
print('Member holonyms:',doggy_synset.member_holonyms())
print('Substance holonyms:',doggy_synset.substance_holonyms())

### Whole - to - part relations
print('Part meronyms:',doggy_synset.part_meronyms())
print('Member meronyms:',doggy_synset.member_meronyms())
print('Substance meronyms:',doggy_synset.substance_meronyms())


Part holonyms: []
Member holonyms: [Synset('canis.n.01'), Synset('pack.n.06')]
Substance holonyms: []
Part meronyms: [Synset('flag.n.07')]
Member meronyms: []
Substance meronyms: []


In [11]:
chase_synset = all_dog_verb_synsets[0]
#### Relations for verbal synsets
print('Caused:', chase_synset.causes())
print('Entailments:',chase_synset.entailments())
print('Hyponyms:', chase_synset.hyponyms())
print('Examples:', chase_synset.examples())


Caused: []
Entailments: []
Hyponyms: [Synset('hound.v.01'), Synset('quest.v.02'), Synset('run_down.v.07'), Synset('tree.v.03')]
Examples: ['The policeman chased the mugger down the alley', 'the dog chased the rabbit']


To get the full Python object definition of a synset to show all options, you can use the 'dir' command:

In [12]:
dir(chase_synset)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__unicode__',
 '__weakref__',
 '_all_hypernyms',
 '_definition',
 '_examples',
 '_frame_ids',
 '_hypernyms',
 '_instance_hypernyms',
 '_iter_hypernym_lists',
 '_lemma_names',
 '_lemma_pointers',
 '_lemmas',
 '_lexname',
 '_max_depth',
 '_min_depth',
 '_name',
 '_needs_root',
 '_offset',
 '_pointers',
 '_pos',
 '_related',
 '_shortest_hypernym_paths',
 '_wordnet_corpus_reader',
 'also_sees',
 'attributes',
 'causes',
 'closure',
 'common_hypernyms',
 'definition',
 'entailments',
 'examples',
 'frame_ids',
 'hypernym_distances',
 'hypernym_paths',
 'hypernyms',
 'hyponyms',
 'in_region_domains',
 'in_topic_domains',
 'in_us

Some of these options may be familair to you if you have read the literature. We will look into the similarity functions more closely later on.

## Wordnets in other languages

There are wordnets in many different languages and many are linked to English. The ones that are freely available in the Open Multilingual Wordnet platform: http://compling.hss.ntu.edu.sg/omw/ are also available in NLTK. You can use "wn.langs" to get the full list.

You might get an error regarding NLTK not finding a 'omw' dataset. You can download it just like you did in lab1.

In [13]:
import nltk
nltk.download('omw')

[nltk_data] Downloading package omw to /Users/piek/nltk_data...
[nltk_data]   Package omw is already up-to-date!


True

Let's check out which languages have wordnets in the OMW package:

In [14]:
sorted(wn.langs())

['als',
 'arb',
 'bul',
 'cat',
 'cmn',
 'dan',
 'ell',
 'eng',
 'eus',
 'fas',
 'fin',
 'fra',
 'glg',
 'heb',
 'hrv',
 'ind',
 'ita',
 'jpn',
 'nld',
 'nno',
 'nob',
 'pol',
 'por',
 'qcn',
 'slv',
 'spa',
 'swe',
 'tha',
 'zsm']

The listed language wordnets are created by translating the English synsets (the so-called Expand Method (Vossen (ed.) 1998). This means that the concepts of the English wordnet are re-used and the synonyms in the synsets are translated.

Since the concept structure is the same for all these wordnets (they share the English concepts), you can  ask for the language lemmas linked to any synset in English.

Are there any Japanese lemmas linked to English dog sense 1

In [16]:
# Are there any Japanese lemmas linked to English dog sense 1
wn.synset('dog.n.01').lemma_names('jpn')

['イヌ', 'ドッグ', '洋犬', '犬', '飼犬', '飼い犬']

In [17]:
# The same for Dutch
wn.synset('dog.n.01').lemma_names('ita')

['cane', 'Canis_familiaris']

So that is great but can we also get the synset directly through a Dutch or Japanese synonym?

In [20]:
wn.synsets('dog.n.01.hond')

[]

Unfortunately not. You cannot directly get the synsets in Wordnet through the same interface we have used before for 'dog'. The next call therefore also does not work:

In [17]:
all_dog_synsets = wn.synsets('hond')
print('Number of synsets with "hond" as a synonym:', len(all_dog_synsets))
print(all_dog_synsets)

Number of synsets with "hond" as a synonym: 0
[]


To obtain the synsets for a non-English word, we first have to use the wn.lemmas() function to get the list of lemma objects for a specific language. The next cell shows this for the Dutch lemma *hond*.

In [18]:
dutch_dog_lemmas = wn.lemmas('hond', lang='nld')
print(dutch_dog_lemmas)

[Lemma('dog.n.01.hond'), Lemma('asshole.n.01.hond')]


In [19]:
type(dutch_dog_lemmas[0])

nltk.corpus.reader.wordnet.Lemma

Lemma is yet another object with attributes and functions, some of which overlap with those of a synset. Let's check them out through 'dir'

In [20]:
dutch_lemma = dutch_dog_lemmas[0]
dir(dutch_lemma)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__slots__',
 '__str__',
 '__subclasshook__',
 '__unicode__',
 '__weakref__',
 '_frame_ids',
 '_frame_strings',
 '_hypernyms',
 '_instance_hypernyms',
 '_key',
 '_lang',
 '_lex_id',
 '_lexname_index',
 '_name',
 '_related',
 '_synset',
 '_syntactic_marker',
 '_wordnet_corpus_reader',
 'also_sees',
 'antonyms',
 'attributes',
 'causes',
 'count',
 'derivationally_related_forms',
 'entailments',
 'frame_ids',
 'frame_strings',
 'hypernyms',
 'hyponyms',
 'in_region_domains',
 'in_topic_domains',
 'in_usage_domains',
 'instance_hypernyms',
 'instance_hyponyms',
 'key',
 'lang',
 'member_holonyms',
 'member_meronyms',
 'name',
 'part_holonyms',
 'part_meronyms',
 'pertainym

Some are different from synset such as lang(). The function *.synset()* can be used to get the synsets associated with a lemma. Obviously, the synset information is the same as for the English wordnet because the Open Dutch Wordnet: http://wordpress.let.vupr.nl/odwn/ was created by expanding the English wordnet.

In [21]:
print(dutch_lemma, dutch_lemma.lang())

dutch_dog_synset = dutch_lemma.synset()
print('Synsets:', dutch_dog_synset)
print('Hypernyms:', dutch_dog_synset.hypernyms())
print('Definition:', dutch_dog_synset.definition())
print('Hyponymy paths:', dutch_dog_synset.hypernym_paths())

Lemma('dog.n.01.hond') nld
Synsets: Synset('dog.n.01')
Hypernyms: [Synset('canine.n.02'), Synset('domestic_animal.n.01')]
Definition: a member of the genus Canis (probably descended from the common wolf) that has been domesticated by man since prehistoric times; occurs in many breeds
Hyponymy paths: [[Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('chordate.n.01'), Synset('vertebrate.n.01'), Synset('mammal.n.01'), Synset('placental.n.01'), Synset('carnivore.n.01'), Synset('canine.n.02'), Synset('dog.n.01')], [Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('domestic_animal.n.01'), Synset('dog.n.01')]]


So we have many wordnets in different languages. Can we get statistics on their coverage?

In [21]:
print('Dutch:', len(list(wn.all_lemma_names(pos='n', lang='nld'))))
print('Italian:', len(list(wn.all_lemma_names(pos='n', lang='ita'))))
print('Japanese:', len(list(wn.all_lemma_names(pos='n', lang='jpn'))))
print('Slovene:', len(list(wn.all_lemma_names(pos='n', lang='slv'))))

print('Spanish:', len(list(wn.all_lemma_names(pos='n', lang='spa'))))

Dutch: 36896
Italian: 31477
Japanese: 64797
Slovene: 31631
Spanish: 28647


### Get the Dutch and Japanese dogs

The next simple "for" loop iterates over all dog-hyponyms in the English wordnet and prints the synset and any Dutch and Japanese labels. We can easily see, which dogs are included in each language and which are not.

In [22]:
dog = wn.synset ('dog.n.01')
dogs = dog.hyponyms()
print('Number of dogs:', len(dogs))

for s in dogs:
    print(s)
    print('Dutch:', s.lemma_names('nld'))
    print('Japanese:', s.lemma_names('jpn'))
    print('Italian:', s.lemma_names('ita'))
    print('Spanish:', s.lemma_names('spa'))
    print()

Number of dogs: 18
Synset('basenji.n.01')
Dutch: []
Japanese: []
Italian: ['basenji', 'cane_del_Congo']
Spanish: ['basenji']

Synset('corgi.n.01')
Dutch: []
Japanese: ['ウェルシュ・コーギー']
Italian: []
Spanish: []

Synset('cur.n.01')
Dutch: ['mormel', 'idioot', 'halve_gare', 'bastaard', 'bastaardhond', 'straathond']
Japanese: ['雑犬', '雑種犬', '駄犬']
Italian: ['bastardo']
Spanish: ['chucho', 'gozque', 'mestizo']

Synset('dalmatian.n.02')
Dutch: ['dalmatiër', 'Dalmatische']
Japanese: []
Italian: ['dalmata']
Spanish: []

Synset('great_pyrenees.n.01')
Dutch: []
Japanese: []
Italian: ['mastino_dei_Pirenei']
Spanish: []

Synset('griffon.n.02')
Dutch: []
Japanese: ['グリフォン', 'ブリュッセルグリフォン', 'グリフォンブリュッセロワ']
Italian: []
Spanish: ['grifón']

Synset('hunting_dog.n.01')
Dutch: []
Japanese: ['猟犬']
Italian: ['cane_da_caccia']
Spanish: []

Synset('lapdog.n.01')
Dutch: []
Japanese: []
Italian: ['cagnolino', 'cagnolino_da_salotto']
Spanish: []

Synset('leonberg.n.01')
Dutch: []
Japanese: []
Italian: []
Spanish: []



This gives dogs as direct hyponyms but maybe there are more dogs as hyponyms of hyponyms of hyponyms, etc.  The WordNet interface documentation uses a so-called anonymous function (lambda) which is applied recursively to synsets that are the hyponyms of synsets. This is higher Python magic. For now accept that it traverses the hyponym tree from a starting synset and puts all results in a list.

In [24]:
hypo = lambda s: s.hyponyms()

In [25]:
dogs_at_all_levels = list(dog.closure(hypo))
print('Number of dogs:', len(dogs_at_all_levels))

Number of dogs: 189


Ahhh, we now have 189 dogs instead of 18 if we go deeper! Let's check their coverage in Dutch

In [26]:

for s in dogs_at_all_levels: 
    print(s)
    print('Dutch:', s.lemma_names('nld'))
    print('Japanese:', s.lemma_names('jpn'))
    print('Italian:', s.lemma_names('ita'))
    print('Spanish:', s.lemma_names('spa'))
    print ()

Synset('basenji.n.01')
Dutch: []
Japanese: []
Italian: ['basenji', 'cane_del_Congo']
Spanish: ['basenji']

Synset('corgi.n.01')
Dutch: []
Japanese: ['ウェルシュ・コーギー']
Italian: []
Spanish: []

Synset('cur.n.01')
Dutch: ['mormel', 'idioot', 'halve_gare', 'bastaard', 'bastaardhond', 'straathond']
Japanese: ['雑犬', '雑種犬', '駄犬']
Italian: ['bastardo']
Spanish: ['chucho', 'gozque', 'mestizo']

Synset('dalmatian.n.02')
Dutch: ['dalmatiër', 'Dalmatische']
Japanese: []
Italian: ['dalmata']
Spanish: []

Synset('great_pyrenees.n.01')
Dutch: []
Japanese: []
Italian: ['mastino_dei_Pirenei']
Spanish: []

Synset('griffon.n.02')
Dutch: []
Japanese: ['グリフォン', 'ブリュッセルグリフォン', 'グリフォンブリュッセロワ']
Italian: []
Spanish: ['grifón']

Synset('hunting_dog.n.01')
Dutch: []
Japanese: ['猟犬']
Italian: ['cane_da_caccia']
Spanish: []

Synset('lapdog.n.01')
Dutch: []
Japanese: []
Italian: ['cagnolino', 'cagnolino_da_salotto']
Spanish: []

Synset('leonberg.n.01')
Dutch: []
Japanese: []
Italian: []
Spanish: []

Synset('mexican_hai

It is clear that the non-English wordnets lack coverage compared to the English WordNet. There is work to do to complete it. Perhaps a nice project for you to work on to increase the coverage of our Dutch WordNet?

The next simple code counts the missing dogs per language.

In [27]:
dog_gaps_nl = 0
dog_gaps_jp = 0
dog_gaps_es = 0
dog_gaps_it = 0


for s in dogs_at_all_levels: 
    if not s.lemma_names('nld'):
        dog_gaps_nl +=1
    
    if not s.lemma_names('jpn'):
        dog_gaps_jp +=1
    
    if not s.lemma_names('ita'):
        dog_gaps_it +=1
    
    if not s.lemma_names('spa'):
        dog_gaps_es +=1
        
print('Missing dogs in Dutch:', dog_gaps_nl)
print('Missing dogs in Japanse:', dog_gaps_jp)
print('Missing dogs in Italian:', dog_gaps_it)
print('Missing dogs in Spanish:', dog_gaps_es)

Missing dogs in Dutch: 142
Missing dogs in Japanse: 100
Missing dogs in Italian: 127
Missing dogs in Spanish: 154


Question: are there any Dutch words for dogs that are not in the English WordNet? Where can you find the answer to such a question?

## Wordnet Similarity

A whole series of similarity functions have been added to NLTK and can be used for scoring synset pairs. See the documentation for the other methods. We show here how it works for "path"

We first obtain a few synsets:

In [25]:
dog = wn.synset('dog.n.01')
cat = wn.synset('cat.n.01')
car = wn.synset('car.n.01')

We can get the full hypernym path for these synsets:

In [40]:
print('dogs are a type of:', dog.hypernym_paths())
print('cats are a type of:', cat.hypernym_paths())

dogs are a type of: [[Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('chordate.n.01'), Synset('vertebrate.n.01'), Synset('mammal.n.01'), Synset('placental.n.01'), Synset('carnivore.n.01'), Synset('canine.n.02'), Synset('dog.n.01')], [Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('domestic_animal.n.01'), Synset('dog.n.01')]]
cats are a type of: [[Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('living_thing.n.01'), Synset('organism.n.01'), Synset('animal.n.01'), Synset('chordate.n.01'), Synset('vertebrate.n.01'), Synset('mammal.n.01'), Synset('placental.n.01'), Synset('carnivore.n.01'), Synset('feline.n.01'), Synset('cat.n.01')]]


We can see that a large parts of the pathes are the same, starting from *carnivore*. Also note that *dog* has two branches one through *domestic_animal* and another one through *canine*.

In [41]:
print('cars are a type of:', car.hypernym_paths())

cars are a type of: [[Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('artifact.n.01'), Synset('instrumentality.n.03'), Synset('container.n.01'), Synset('wheeled_vehicle.n.01'), Synset('self-propelled_vehicle.n.01'), Synset('motor_vehicle.n.01'), Synset('car.n.01')], [Synset('entity.n.01'), Synset('physical_entity.n.01'), Synset('object.n.01'), Synset('whole.n.02'), Synset('artifact.n.01'), Synset('instrumentality.n.03'), Synset('conveyance.n.03'), Synset('vehicle.n.01'), Synset('wheeled_vehicle.n.01'), Synset('self-propelled_vehicle.n.01'), Synset('motor_vehicle.n.01'), Synset('car.n.01')]]


Obviously, the path for *car* is very different but at some point meets the other pathes at *whole.n.02*.

Wordnet similarity measures compare the distance from word meanings through the hypernym path and possibly other relations.

Synsets have the similarity function built in. The function requires another synset as the input and gives back a score:

In [37]:
print(dog.path_similarity(cat))

0.2


Is this very similar? The only way to find out is to compare this with others:

In [42]:
print(car.path_similarity(cat))
print(car.path_similarity(dog))

0.05555555555555555
0.07692307692307693


Right, these score a lot lower.

Let's see of this also works for verbs.

In [43]:
hit = wn.synset('hit.v.01')
slap = wn.synset('slap.v.01')
print(hit.path_similarity(slap))
print(wn.path_similarity(hit, slap))

0.14285714285714285
0.14285714285714285


It seems to work as well.

If you did some readings on WordNet, you may know that the noun hierarchy has a single top-node synset 'entity-n-01'. All nominal synsets decent from this synset. This is not the case for verbs nor for adjectives. The verb part of WordNet therefore consists of '559' islands of disconnected synsets with 559 rootnodes. The English WordNet editors decided not to connect these islands in an artificial way as was done for nouns.

In [44]:
print('hit is a type of:', hit.hypernym_paths())
print('slap is a type of:', slap.hypernym_paths())

hit is a type of: [[Synset('move.v.02'), Synset('propel.v.01'), Synset('hit.v.01')]]
slap is a type of: [[Synset('touch.v.01'), Synset('strike.v.01'), Synset('slap.v.01')]]


We see there is no overlap between the pathes for the two verbs.

We can also see this using the NLTK root_hypernyms() function for the above noun and verb synsets.

In [45]:
print('Root for hit:', dog.root_hypernyms())
print('Root for hit:', cat.root_hypernyms())
print('Root for hit:', slap.root_hypernyms())
print('Root for hit:', hit.root_hypernyms())

Root for hit: [Synset('entity.n.01')]
Root for hit: [Synset('entity.n.01')]
Root for hit: [Synset('touch.v.01')]
Root for hit: [Synset('move.v.02')]


How is it still possible to get a value for similarity if the subgraphs are not connected? Well, the package imposes a simulated rootnode by grouping all the subgraph top-nodes under a single node. This is the default setting. If you do not want to use this, you can turn it off by setting the parameter *simulate_root* to False.

In [46]:
print(hit.path_similarity(slap, simulate_root=False))
print(wn.path_similarity(hit, slap, simulate_root=False))

None
None


In [26]:
print(cat.path_similarity(cat))



1.0


Without the simulated root there is no path from 'hit' to 'slap'.

## Using WordNet similarity for words instead of synsets

Now if we want to use this for words, we first need to obtain all the synsets for a word and then compare each synset with the synsets of another word. We thus need a for-loop inside a for-loop. The first loop gets the synsets for the first word and the second loop for each synset the synsets for the second word to compare.

In [32]:
w1='dog'
w2='cat'
for s1 in wn.synsets(w1, 'n'):
    print(s1,':')
    for s2 in wn.synsets(w2, 'n'):
        print('\t', s2,':', s1.path_similarity(s2))

Synset('dog.n.01') :
	 Synset('cat.n.01') : 0.2
	 Synset('guy.n.01') : 0.125
	 Synset('cat.n.03') : 0.125
	 Synset('kat.n.01') : 0.07692307692307693
	 Synset('cat-o'-nine-tails.n.01') : 0.08333333333333333
	 Synset('caterpillar.n.02') : 0.07692307692307693
	 Synset('big_cat.n.01') : 0.2
	 Synset('computerized_tomography.n.01') : 0.05263157894736842
Synset('frump.n.01') :
	 Synset('cat.n.01') : 0.07142857142857142
	 Synset('guy.n.01') : 0.125
	 Synset('cat.n.03') : 0.125
	 Synset('kat.n.01') : 0.1
	 Synset('cat-o'-nine-tails.n.01') : 0.07142857142857142
	 Synset('caterpillar.n.02') : 0.06666666666666667
	 Synset('big_cat.n.01') : 0.07142857142857142
	 Synset('computerized_tomography.n.01') : 0.05555555555555555
Synset('dog.n.03') :
	 Synset('cat.n.01') : 0.07692307692307693
	 Synset('guy.n.01') : 0.2
	 Synset('cat.n.03') : 0.14285714285714285
	 Synset('kat.n.01') : 0.1111111111111111
	 Synset('cat-o'-nine-tails.n.01') : 0.07692307692307693
	 Synset('caterpillar.n.02') : 0.07142857142857

We can use the highest similarity from all pairs to find the strongest association.

In [54]:
w1='dog'
w2='cat'
for s1 in wn.synsets(w1, 'n'):
    top_sim_score = 0    
    top_sim_synset_w1 = ""
    top_sim_synset_w2 = ""
    for s2 in wn.synsets(w2, 'n'):
        sim = s1.path_similarity(s2)
        if sim>top_sim_score:
            top_sim_score = sim
            top_sim_synset_w2 = s2
    print('Most similar are', s1, top_sim_synset_w2,':', top_sim_score)

Most similar are Synset('dog.n.01') Synset('cat.n.01') : 0.2
Most similar are Synset('frump.n.01') Synset('guy.n.01') : 0.125
Most similar are Synset('dog.n.03') Synset('guy.n.01') : 0.2
Most similar are Synset('cad.n.01') Synset('guy.n.01') : 0.14285714285714285
Most similar are Synset('frank.n.02') Synset('kat.n.01') : 0.09090909090909091
Most similar are Synset('pawl.n.01') Synset('cat-o'-nine-tails.n.01') : 0.14285714285714285
Most similar are Synset('andiron.n.01') Synset('cat-o'-nine-tails.n.01') : 0.16666666666666666


# End of this notebook